src/common/subprocess/subprocess.h

```diff
@@ -0,0 +1,370 @@
+                                   * all other errors */
+} flux_subprocess_state_t;
+
+/*
+ * Subprocess flags
+ */
+enum {
+    /* flux_exec(): let parent stdin, stdout, stderr, carry to child.
+     * Do not create "STDIN", "STDOUT", or "STDERR" channels.  Subsequently,
+     * flux_subprocess_write()/close()/read()/read_line() will fail on
+     * streams of "STDIN", "STDOUT", or "STDERR".
+     */
+    FLUX_SUBPROCESS_FLAGS_STDIO_FALLTHROUGH = 1,
+    /* flux_exec(): call setpgrp() before exec(2) */
+    FLUX_SUBPROCESS_FLAGS_SETPGRP = 2,
+};
+
+/*
+ *  Typedefs for subprocess hooks and callbacks:
+ *
+ */
+typedef void (*flux_subprocess_f) (flux_subprocess_t *p);
+typedef void (*flux_subprocess_output_f) (flux_subprocess_t *p,
+                                          const char *stream);
+typedef void (*flux_subprocess_state_f) (flux_subprocess_t *p,
+                                         flux_subprocess_state_t state);
+
+/*
+ *  Functions for event-driven subprocess handling:
+ *
+ */
+typedef struct {
+    flux_subprocess_f on_completion;    /* Process exited and all I/O
+                                         * complete, will not be
+                                         * called if EXEC_FAILED or
+                                         * FAILED states reached.
+                                         */
+    flux_subprocess_state_f on_state_change;  /* Process state change          */
+    flux_subprocess_output_f on_channel_out; /* Read from channel when ready   */
+    flux_subprocess_output_f on_stdout; /* Read of stdout is ready             */
+    flux_subprocess_output_f on_stderr; /* Read of stderr is ready             */
+} flux_subprocess_ops_t;
+
+/*
+ *  General support:
+ */
+
+/*  Start a subprocess server on the handle `h`. Registers message
+ *   handlers, etc for remote execution. "prefix" is the topic prefix
+ *   used to listen for this service, e.g. `broker` would listen
+ *   for `broker.exec`.
+ */
+flux_subprocess_server_t *flux_subprocess_server_start (flux_t *h,
+                                                        const char *prefix,
+                                                        const char *local_uri,
+                                                        uint32_t rank);
+
+/*  Stop a subprocess server / cleanup flux_subprocess_server_t */
+void flux_subprocess_server_stop (flux_subprocess_server_t *s);
+
+/* Terminate all subprocesses started by a sender id */
+int flux_subprocess_server_terminate_by_uuid (flux_subprocess_server_t *s,
+                                              const char *id);
+
+/*  General output callback that will send output from the subprocess
+ *  to stdout or stderr.  Set to `on_stdout` and/or `on_stderr` in
+ *  flux_subprocess_ops_t.  Can also be used for 'on_channel_out'
+ *  callback, sending all output to stdout.
+ */
+void flux_subprocess_output (flux_subprocess_t *p, const char *stream);
```

Comment:
The comment here or perhaps the name of the function confused me here (could just be my fault).
`flux_subprocess_output` is called a "callback" but it isn't a typedef like you'd see in the case of a normal callback. Is this function actually just a convenience function that can be used in an output *reader* function to send output to the standard `stdout` stream of subprocess `p`?

Does this function actually need to be in the public header? (Sorry I'm probably just misunderstanding from cursory view)

Reply:
Yeah, this is just a convenience function.  If all the user desires is to send the subprocesses' stdout/stderr data to the parent's stdout/stderr, this is a "callback" that you an set in ```on_stdout``` or ```on_stderr```.  Perhaps the name of the function is bad.  Minimally, it should not be prefixed as ```flux_subprocess```.  Perhaps ```flux_dump()``` or something similar?

Reply:
Yeah, I'm still a bit confused, and I'm sure it is just my fault.

Could sending output to the parent stdout/err just be the default? Another thing that throws me about this being a "callback" is that there is no "data" argument...

Since the first argument is a `flux_subprocess_t`, I don't have a problem with it being a `flux_subprocess_*` method. But I'm not sure at this point how to make it less confusing...

Reply:
> Could sending output to the parent stdout/err just be the default?

I'll ponder this.  I now realize that I semi-associated "default dump to parent stdout/stderr"  with "inherit parent stdout/stderr".  And now realize those are different things and I can separate that from each other.

Reply:
Thinking about this more, having the default stdout/stderr output goto to the parent's stdout/stderr is the right move.  If anything, it's the more common case.  We can then  remove the ```flux_subprocess_output()``` function.